### PR TITLE
AnaTuple production

### PR DESCRIPTION
**Pull request by valeriadamante**
This PR is intended to fix the anatuple production code and to make histograms

### CODE REVIEW

File: AnaProd/tasks.py [hunk 1]
```diff
@@ -116,4 +161,80 @@ def getBranches(samples, central_nanoAOD_path, max_files_per_job):
...
+            split_idx = 0
+            while True:
+                start_idx, stop_idx = split_idx * max_files_per_job, (split_idx + 1) * max_files_per_job
+                branches[n] = (sample_id, sample_name, samples[sample_name]['sampleType'], split_idx,
+                               input_files[start_idx:stop_idx])
+                split_idx += 1
+                n += 1
+                if stop_idx >= len(input_files):
+                    break
         return branches
```
Comment (kandrosov) [inline AnaProd/tasks.py:240]: reimplement
- there is only one branch for merging data
- don't rely on input files to define branch_data, just list of sample names

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: AnaProd/MergeNtuples.py [hunk 2]
```diff
@@ -19,5 +31,91 @@
...
+            key_name = k.GetName()
+            if k.GetClassName() =='TTree':
+                if key_name not in dfNames:
+                    dfNames.append(k.GetName())
+            else:
+                obj = k.ReadObj()
+                if isinstance(obj, ROOT.TH1):
+                    if key_name not in histNames:
+                        histNames.append(k.GetName())
+                        histograms[key_name] = ROOT.TH1D()
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/MergeNtuples.py:75]: already tried. Clone does not work

File: AnaProd/tasks.py [hunk 3]
```diff
@@ -59,34 +59,50 @@ def requires(self):
...
+        if sample_type=='data':
+            tmpFile = os.path.join(outdir_anatuples, outFileName)
+        outdir_final = os.path.join(self.central_anaTuples_path(), sample_name)
+        os.makedirs(outdir_final, exist_ok=True)
+        finalFile = self.output().path
+        shutil.copy(tmpFile, finalFile)
+        if os.path.exists(finalFile):
+            os.remove(tmpFile)
+        else:
+            print(f"error, the file {tmpFile} was not copied in {finalFile} ")
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/tasks.py:98]: ok

File: AnaProd/tasks.py [hunk 4]
```diff
@@ -95,16 +111,45 @@ def getInputFiles(central_nanoAOD_path, sample_name):
...
-    def getBranches(samples, central_nanoAOD_path, max_files_per_job):
+    def getBranches(samples, central_nanoAOD_path, max_files_per_job, txtFilePath):
         n = 0
         branches = {}
         for sample_id, sample_name in enumerate(sorted(samples.keys())):
-            input_files = AnaTupleTask.getInputFiles(central_nanoAOD_path, sample_name)
+            txtFilePath_complete = os.path.join(txtFilePath,sample_name)
+            os.makedirs(txtFilePath_complete, exist_ok=True)
+            txtFile = os.path.join(txtFilePath_complete, 'input_files.txt')
+            if os.path.exists(txtFile):
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/tasks.py:146]: ah ok

File: AnaProd/tasks.py [hunk 5]
```diff
@@ -116,4 +161,80 @@ def getBranches(samples, central_nanoAOD_path, max_files_per_job):
...
                 if stop_idx >= len(input_files):
                     break
+            if os.path.exists(txtFile):
+                os.remove(txtFile)
+        return branches
+
+
+class DataMergingTask(Task, HTCondorWorkflow, law.LocalWorkflow):
+    max_runtime = copy_param(HTCondorWorkflow.max_runtime, 2.0)
+    max_files_per_job = 100000 #luigi.IntParameter(default=1, description="maximum number of input files per job")
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/tasks.py:171]: ok

File: AnaProd/tasks.py [hunk 6]
```diff
@@ -116,4 +161,80 @@ def getBranches(samples, central_nanoAOD_path, max_files_per_job):
...
+
+    def law_job_home(self):
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/tasks.py:173]: ok

File: AnaProd/MergeNtuples.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kandrosov) [reply AnaProd/MergeNtuples.py:75]: try obj.Merge as [here](https://github.com/kandrosov/RunKit/blob/ebe3d5ce2432d2620d616df1f4228aa43ac5eb46/haddnano.py#L108C10-L108C10)

File: AnaProd/tasks.py [hunk 7]
```diff
@@ -116,4 +161,80 @@ def getBranches(samples, central_nanoAOD_path, max_files_per_job):
...
+
+    @staticmethod
+    def getInputFiles(central_anatuple_path, sample_name):
+        inDir = os.path.join(central_anatuple_path, sample_name)
+        input_files = []
+        for root, dirs, files in os.walk(inDir):
+            for file in files:
+                if file.endswith('.root') and not file.startswith('.'):
+                    input_files.append(os.path.join(root, file))
+        return list(sorted(input_files))
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/tasks.py]: ok

File: AnaProd/tasks.py [hunk 8]
```diff
@@ -40,71 +39,125 @@ def run(self):
...
+            input_lines = inputFileTxt.read().splitlines()
+            for input_line in input_files:
+                if input_line + '\n' in input_lines:
+                    continue
+                input_lines.append(input_line + '\n')
+        with open(txtFile_tmp, 'a') as inputtxtFile:
+            inputtxtFile.writelines(input_lines)
+        finalFile = self.output().path
+        if os.path.exists(txtFile_tmp):
+            shutil.move(txtFile_tmp,finalFile)
```
Comment (kandrosov) [inline AnaProd/tasks.py:87]:
```suggestion
        shutil.move(txtFile_tmp,finalFile)
```

File: AnaProd/tasks.py [hunk 9]
```diff
@@ -40,71 +39,125 @@ def run(self):
...
+
+    def run(self):
+        sample_name = self.branch_data
+        print(f'Creating anaCache for sample {sample_name} into {self.output().path}')
+        os.makedirs(os.path.join(self.local_path(),sample_name), exist_ok=True)
+        txtFile_tmp = os.path.join(self.local_path(), sample_name, "tmp.txt")
+        if os.path.exists(txtFile_tmp):
+            os.remove(txtFile_tmp)
+        with open(txtFile_tmp, 'a'):
+            os.utime(txtFile_tmp)
```
Comment (kandrosov) [inline AnaProd/tasks.py:69]:
```suggestion
```

File: AnaProd/tasks.py [hunk 10]
```diff
@@ -40,71 +39,125 @@ def run(self):
...
+        n = 0
+        branches = {}
+        for sample_name in sorted(self.samples.keys()):
+            branches[n] = sample_name
+            n += 1
+        return branches
+
+    def output(self):
+        sample_name = self.branch_data
+        sample_out = os.path.join(self.central_anaCache_path(), sample_name, 'input_files.txt')
```
Comment (kandrosov) [inline AnaProd/tasks.py:58]: don't use central path on EOS

File: AnaProd/tasks.py [hunk 11]
```diff
@@ -40,71 +39,125 @@ def run(self):
...
+        with open(txtFile_tmp, 'a'):
+            os.utime(txtFile_tmp)
+        inDir = os.path.join(self.central_nanoAOD_path(), sample_name)
+        input_files = []
+        for root, dirs, files in os.walk(inDir):
+            for file in files:
+                if file.endswith('.root') and not file.startswith('.'):
+                    if os.path.join(root, file) not in input_files:
+                        input_files.append(os.path.join(root, file))
+        with open(txtFile_tmp, 'r') as inputFileTxt:
```
Comment (kandrosov) [inline AnaProd/tasks.py:77]:
```suggestion
        with open(txtFile_tmp, 'w') as inputFileTxt:
```

File: AnaProd/tasks.py [hunk 12]
```diff
@@ -40,71 +39,125 @@ def run(self):
...
+            input_lines = inputFileTxt.read().splitlines()
```
Comment (kandrosov) [inline AnaProd/tasks.py:78]:
```suggestion
```

File: AnaProd/tasks.py [hunk 13]
```diff
@@ -40,71 +39,125 @@ def run(self):
...
+            for input_line in input_files:
+                if input_line + '\n' in input_lines:
+                    continue
+                input_lines.append(input_line + '\n')
```
Comment (kandrosov) [inline AnaProd/tasks.py:82]:
```suggestion
inputFileTxt.write(input_line)
```

File: AnaProd/tasks.py [hunk 14]
```diff
@@ -40,71 +39,125 @@ def run(self):
...
+        with open(txtFile_tmp, 'a') as inputtxtFile:
+            inputtxtFile.writelines(input_lines)
```
Comment (kandrosov) [inline AnaProd/tasks.py:84]:
```suggestion
```

File: Analysis/Producer.py [hunk 15]
```diff
@@ -118,7 +118,7 @@ def FixNegativeContributions(histogram):
     RenormalizeHistogram(histogram, original_Integral, True)
     return True, ss_debug, ss_negative
 
-def GetValues(collection):
+def GetValues(collectio):
```
Comment (kandrosov) [inline Analysis/Producer.py:121]: undo

File: Analysis/SkimProducer.py [hunk 16]
```diff
@@ -2,13 +2,26 @@
...
-
+ROOT.EnableThreadSafety()
 if __name__ == "__main__":
     sys.path.append(os.environ['ANALYSIS_PATH'])
 
 from RunKit.sh_tools import sh_call
 import Common.ConvertUproot as ConvertUproot
 
+
+def getTreeName(systFile):
```
Comment (kandrosov) [inline Analysis/SkimProducer.py:13]: why it is needed?

File: Common/ConvertUproot.py [hunk 17]
```diff
@@ -1,44 +1,66 @@
...
-    with uproot.recreate(outFile, compression=uproot.LZMA(9)) as out_file:
-      out_file[treeName] = out_tree
-      out_file.close()
-    return outFile
+  with uproot.recreate(outFile, compression=uproot.LZMA(9)) as out_file:
+    for hist in histograms.keys():
+      out_file[hist] = histograms[hist]
+    for out_tree_key in out_tree_dict.keys():
+      out_file[out_tree_key] = out_tree_dict[out_tree_key]
+  out_file.close()
```
Comment (kandrosov) [inline Common/ConvertUproot.py:16]: You are using `with` above. `close` should be redundant.

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: AnaProd/tasks.py (hunk 8 above)
Comment (valeriadamante) [reply AnaProd/tasks.py:87]: ok

File: AnaProd/tasks.py (hunk 10 above)
Comment (valeriadamante) [reply AnaProd/tasks.py:58]: ok, I use:
`self.local_path(sample_name, "input_files.txt")`

File: AnaProd/tasks.py (hunk 11 above)
Comment (valeriadamante) [reply AnaProd/tasks.py:77]: ah ok

File: AnaProd/tasks.py (hunk 12 above)
Comment (valeriadamante) [reply AnaProd/tasks.py:78]: ok, I create `input_lines = [] ` above

File: Analysis/Producer.py (hunk 15 above)
Comment (valeriadamante) [reply Analysis/Producer.py:121]: ok

File: Analysis/SkimProducer.py (hunk 16 above)
Comment (valeriadamante) [reply Analysis/SkimProducer.py:13]: to create a proper tree name

File: Common/ConvertUproot.py (hunk 17 above)
Comment (valeriadamante) [reply Common/ConvertUproot.py:16]: ok

File: AnaProd/tasks.py [hunk 18]
```diff
@@ -40,71 +39,117 @@ def run(self):
...
+        input_files = []
+        for root, dirs, files in os.walk(inDir):
+            for file in files:
+                if file.endswith('.root') and not file.startswith('.'):
+                    if os.path.join(root, file) not in input_files:
+                        input_files.append(os.path.join(root, file))
+        input_lines = []
+        with open(txtFile_tmp, 'w') as inputFileTxt:
+
+            #input_lines = inputFileTxt.read().splitlines()
```
Comment (kandrosov) [inline AnaProd/tasks.py:76]:
```suggestion
        with open(txtFile_tmp, 'w') as inputFileTxt:
```

File: AnaProd/tasks.py [hunk 19]
```diff
@@ -40,71 +39,117 @@ def run(self):
...
 
     def requires(self):
         sample_id, sample_name, sample_type, split_idx, input_files = self.branch_data
-        return [ AnaCacheTask.req(self, branch=sample_id, max_runtime=AnaCacheTask.max_runtime._default) ]
+        return [ AnaCacheTask.req(self, branch=sample_id, max_runtime=AnaCacheTask.max_runtime._default), InputFileTask.req(self, branch=sample_id, workflow='local') ]
 
     def create_branch_map(self):
         self.load_sample_configs()
-        return AnaTupleTask.getBranches(self.samples, self.central_nanoAOD_path(), self.max_files_per_job)
+        return AnaTupleTask.getBranches(self.samples, self.max_files_per_job, self.central_anaCache_path())
```
Comment (kandrosov) [inline AnaProd/tasks.py:99]: The current code doesn't work if `max_files_per_job` != 1. Please, remove this parameter and update the code (e.g. input_files -> input_file etc.)

File: Common/ConvertUproot.py [hunk 20]
```diff
@@ -1,44 +1,65 @@
...
   if os.path.exists(outFile):
     os.remove(outFile)
-    with uproot.recreate(outFile, compression=uproot.LZMA(9)) as out_file:
-      out_file[treeName] = out_tree
-      out_file.close()
-    return outFile
+  with uproot.recreate(outFile, compression=uproot.LZMA(9)) as out_file:
+    for hist in histograms.keys():
+      out_file[hist] = histograms[hist]
+    for out_tree_key in out_tree_dict.keys():
```
Comment (kandrosov) [inline Common/ConvertUproot.py:14]: Why do we need a multi-tree version?

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: AnaProd/tasks.py [hunk 21]
```diff
@@ -40,71 +39,117 @@ def run(self):
...
+class InputFileTask(Task, law.LocalWorkflow):
+    max_files_per_job = luigi.IntParameter(default=1, description="maximum number of input files per job")
+
+    def create_branch_map(self):
+        self.load_sample_configs()
+        n = 0
+        branches = {}
+        for sample_name in sorted(self.samples.keys()):
+            branches[n] = sample_name
+            n += 1
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/tasks.py:53]: ok

File: AnaProd/tasks.py (hunk 18 above)
Comment (valeriadamante) [reply AnaProd/tasks.py:76]: ok

File: Common/ConvertUproot.py (hunk 20 above)
Comment (valeriadamante) [reply Common/ConvertUproot.py:14]: [conversation on telegram]

File: AnaProd/tasks.py [hunk 22]
```diff
@@ -40,71 +39,117 @@ def run(self):
...
+        return branches
+
+    def output(self):
+        sample_name = self.branch_data
+        sample_out = os.path.join(self.local_path(sample_name, "input_files.txt"))
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/tasks.py:58]: wait, I added to store_parts but this line shall I just leave as-it-is?

File: AnaProd/tasks.py (hunk 19 above)
Comment (valeriadamante) [reply AnaProd/tasks.py:99]: ok

File: AnaProd/tasks.py [hunk 23]
```diff
@@ -40,80 +39,123 @@ def run(self):
...
+        os.makedirs(os.path.join(self.local_path(),sample_name), exist_ok=True)
+        txtFile_tmp = os.path.join(self.local_path(), sample_name, "tmp.txt")
+        inDir = os.path.join(self.central_nanoAOD_path(), sample_name)
+        input_files = []
+        for root, dirs, files in os.walk(inDir):
+            for file in files:
+                if file.endswith('.root') and not file.startswith('.'):
+                    if os.path.join(root, file) not in input_files:
+                        input_files.append(os.path.join(root, file))
+        input_lines = []
```
Comment (kandrosov) [inline AnaProd/tasks.py:71]:
```suggestion
```

File: AnaProd/tasks.py [hunk 24]
```diff
@@ -40,80 +39,123 @@ def run(self):
...
         n = 0
         branches = {}
         for sample_id, sample_name in enumerate(sorted(samples.keys())):
-            input_files = AnaTupleTask.getInputFiles(central_nanoAOD_path, sample_name)
+            inputFileTxt = os.path.join(central_anacache_path, sample_name, 'input_files.txt')
+            with open(inputFileTxt, 'r') as inputtxtFile:
+                input_files = inputtxtFile.read().splitlines()
             if len(input_files) == 0:
                 raise RuntimeError(f"AnaTupleTask: no input files found for {sample_name}")
             split_idx = 0
```
Comment (kandrosov) [inline AnaProd/tasks.py:151]:
```
for input_file in input_files:
    branches[n] = (sample_id, sample_name, samples[sample_name]['sampleType'], input_file)
    n += 1
```

File: AnaProd/tasks.py [hunk 25]
```diff
@@ -40,80 +39,123 @@ def run(self):
...
-                  self.customisations, '--compute_unc_variations', 'True', '--store-noncentral', '--nEvents', '100'], env=self.cmssw_env())
+                  #self.customisations, '--compute_unc_variations', 'True', '--store-noncentral', '--nEvents', '10'], env=self.cmssw_env(),verbose=1)
+                  self.customisations, '--compute_unc_variations', 'True', '--store-noncentral'], env=self.cmssw_env(),verbose=1)
         producer_skimtuples = os.path.join(self.ana_path(), 'Analysis', 'SkimProducer.py')
         outdir_skimtuples = os.path.join(job_home, 'skim', sample_name)
-        sh_call([ 'python3', producer_skimtuples, '--inputDir',outdir_anatuples, '--workingDir', outdir_skimtuples, '--outputFile', 'skim.root'],verbose=1)
-        outdir_final = self.output().path
-        print(os.listdir(outdir_skimtuples))
-        shutil.move(outdir_skimtuples, outdir_final)
+        outFileName = input_file[0].split('/')[-1]
```
Comment (kandrosov) [inline AnaProd/tasks.py:116]:
```suggestion
        outFileName = os.path.basename(input_file)
```

File: run_tools/law_customizations.py [hunk 26]
```diff
@@ -92,7 +93,7 @@ def load_sample_configs(self):
         self.samples = { key : samples[key] for key in selected_samples }
 
     def store_parts(self):
-        return (self.__class__.__name__, self.version)
+        return (self.__class__.__name__, self.version, self.period)
```
Comment (kandrosov) [inline run_tools/law_customizations.py:96]:
```suggestion
        return (self.__class__.__name__, self.period, self.version)
```
to be compatible with `central_anaTuples_path`

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: run_tools/law_customizations.py (hunk 26 above)
Comment (valeriadamante) [reply run_tools/law_customizations.py:96]: ok

File: AnaProd/tasks.py (hunk 25 above)
Comment (valeriadamante) [reply AnaProd/tasks.py:116]: ok

File: AnaProd/tasks.py [hunk 27]
```diff
@@ -40,80 +39,123 @@ def run(self):
...
+
 
     def workflow_requires(self):
-        return { "anaCache" : AnaCacheTask.req(self) }
+        return { "anaCache" : AnaCacheTask.req(self), "inputFile": InputFileTask.req(self,workflow='local') }
 
     def requires(self):
-        sample_id, sample_name, sample_type, split_idx, input_files = self.branch_data
-        return [ AnaCacheTask.req(self, branch=sample_id, max_runtime=AnaCacheTask.max_runtime._default) ]
+        sample_id, sample_name, sample_type, split_idx, input_file = self.branch_data
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/tasks.py:90]: ok

File: AnaProd/tasks.py (hunk 23 above)
Comment (valeriadamante) [reply AnaProd/tasks.py:71]: ok

File: AnaProd/tasks.py [hunk 28]
```diff
@@ -40,80 +39,124 @@ def run(self):
...
-                start_idx, stop_idx = split_idx * max_files_per_job, (split_idx + 1) * max_files_per_job
-                branches[n] = (sample_id, sample_name, samples[sample_name]['sampleType'], split_idx,
-                               input_files[start_idx:stop_idx])
-                split_idx += 1
+            for input_file in input_files:
+                branches[n] = (sample_id, sample_name, samples[sample_name]['sampleType'], input_file)
                 n += 1
-                if stop_idx >= len(input_files):
+                if n >= len(input_files):
                     break
```
Comment (kandrosov) [inline AnaProd/tasks.py:160]:
```suggestion
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: AnaProd/tasks.py (hunk 28 above)
Comment (valeriadamante) [reply AnaProd/tasks.py:160]: ok

**Review by kandrosov — APPROVED** [review]

**Pull request merged**